File: auction/auction.go
package auction

type Attendees int

const (
	AtA Attendees = iota
	AtB
)

type Auctioner interface {
	Do() (Attendees, int)
}

type auctionerImpl struct {
	initPrice int
	aLimit    int
	bLimit    int
}

func New(initPrice, aLimit, bLimit int) Auctioner {
	return &auctionerImpl{initPrice, aLimit, bLimit}
}

func (a *auctionerImpl) Do() (Attendees, int) {
	price := a.initPrice
	for i := 0; ; i++ {
		if i%2 == 0 {
			// a turn
			nextPrice := price + 10
			if nextPrice < a.aLimit {
				price = nextPrice
			} else {
				return AtB, price
			}
		} else {
			// b turn
			nextPrice := price + 1000
			if nextPrice < a.bLimit {
				price = nextPrice
			} else {
				return AtA, price
			}
		}
	}
}

File: src/auction.rs
#[derive(Debug, PartialEq)]
enum Attendee {
    AtA,
    AtB,
}

trait Auctioner {
    fn execute(&self) -> (Attendee, u32);
}

struct AuctionerImpl {
    init_price: u32,
    a_limit: u32,
    b_limit: u32,
}

impl AuctionerImpl{
    fn new(init_price:u32, a_limit:u32, b_limit:u32) -> AuctionerImpl {
        AuctionerImpl{init_price, a_limit, b_limit}
    }
}

fn new(init_price:u32, a_limit:u32, b_limit:u32) -> impl Auctioner {
    AuctionerImpl::new(init_price, a_limit, b_limit)
}

impl Auctioner for AuctionerImpl {
    fn execute(&self) -> (Attendee, u32){
        let mut price = self.init_price;
        let mut i = 0;
        loop {
            if i % 2 == 0 {
                let next_price = price + 10;
                if next_price < self.a_limit {
                    price = next_price;
                } else {
                    return (Attendee::AtB, price);
                }
            } else {
                let next_price = price + 1000;
                if next_price < self.b_limit {
                    price = next_price;
                } else {
                    return (Attendee::AtA, price)
                }
            }
            i += 1;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Auctioner;
    #[test]
    fn test(){
        let auctioner = super::new(1, 1500, 2050);
        let (winner, price) = auctioner.execute();
        assert_eq!(winner, super::Attendee::AtB);
        assert_eq!(price, 2021);
    }
}
